module/apmstreadwayamqp/pub.go

```diff
@@ -0,0 +1,72 @@
+package apmstreadwayamqp
+
+import (
+	"context"
+	"github.com/streadway/amqp"
+	"go.elastic.co/apm/v2"
+)
+
+// Publisher is the interface returned by ContextConn.
+//
+// Publisher's Publish method reports spans using the bound context.
+type Publisher interface {
+
+	// WithContext returns a shallow copy of the Publisher with
+	// its context changed to ctx.
+	//
+	// To report publishing as spans, ctx must contain a transaction or span.
+	WithContext(ctx context.Context) Publisher
+
+	// Publish publishes a message and returns an error in encountered.
+	//
+	// The started span is stored in the ctx.
+	Publish(exchange, key string, mandatory, immediate bool, msg amqp.Publishing) error
+}
+
+// Wrap wraps amqp.Channel such that its Publish method calls streadway/amqp.Publish.
+func Wrap(ch *amqp.Channel) Publisher {
+	return msgPublisher{ch: ch, ctx: context.Background()}
+}
+
+type msgPublisher struct {
+	ch  *amqp.Channel
+	ctx context.Context
+}
+
+func (c msgPublisher) WithContext(ctx context.Context) Publisher {
+	c.ctx = ctx
+	return c
+}
+
+func (c msgPublisher) Publish(exchange, key string, mandatory, immediate bool, msg amqp.Publishing) error {
+	return pub(c.ctx, c.ch, exchange, key, mandatory, immediate, msg)
+}
+
+func pub(ctx context.Context, ch *amqp.Channel, exchange, key string, mandatory, immediate bool, msg amqp.Publishing) error {
+	span, ctx := apm.StartSpanOptions(ctx, "rabbitmq.pub", "messaging", apm.SpanOptions{
+		ExitSpan: true,
+	})
+	span.Subtype = "rabbitmq"
+	span.Outcome = "success"
+	defer span.End()
+
+	setMessageTraceparent(span.TraceContext(), msg.Headers)
+	stateErr := setMessageTracestate(span.TraceContext().State, msg.Headers)
```

Comment:
If we introduce a package-level `ExtractTraceContext` function, perhaps we should also extract these into a `InjectTraceContext` function:

```go
func InjectTraceContext(apm.TraceContext, amqp.Delivery)
```

Reply:
Also, in some cases you should use the transaction context rather than the span context. Specifically, when the span is dropped due to the max spans per transaction having been reached. In this case the returned span won't have a trace context.

Here's how the apmhttp client instrumentation does it, for reference: https://github.com/elastic/apm-agent-go/blob/acf7d2b497fa70d5a1bb06088f700c152de5fb0c/module/apmhttp/client.go#L85-L120